src/ucs/async/thread.c

```diff
@@ -90,31 +113,13 @@ static void *ucs_async_thread_func(void *arg)
             timeout_ms = ucs_time_to_msec(timer_interval -
                                           ucs_min(time_spent, timer_interval));
         }
-        nready = epoll_wait(thread->epfd, events, UCS_ASYNC_EPOLL_MAX_EVENTS,
-                            timeout_ms);
-        if ((nready < 0) && (errno != EINTR)) {
-            ucs_fatal("epoll_wait() failed: %m");
-        }
-        ucs_trace_async("epoll_wait(epfd=%d, timeout=%d) returned %d",
-                        thread->epfd, timeout_ms, nready);
-
-        /* Check ready files */
-        if (nready > 0) {
-            for (i = 0; i < nready; ++i) {
-                fd = events[i].data.fd;
-
-                /* Check wakeup pipe */
-                if (fd == ucs_async_pipe_rfd(&thread->wakeup)) {
-                    ucs_trace_async("progress thread woken up");
-                    ucs_async_pipe_drain(&thread->wakeup);
-                    continue;
-                }
-
-                status = ucs_async_dispatch_handlers(&fd, 1);
-                if (status == UCS_ERR_NO_PROGRESS) {
-                    is_missed = 1;
-                }
-            }
+
+        status = ucs_event_set_wait(thread->event_set,
+                                    UCS_ASYNC_EPOLL_MAX_EVENTS, timeout_ms,
+                                    ucs_async_thread_ev_handler, cb_arg,
+                                    &read_events);
+        if (status != UCS_OK) {
+            ucs_error("ucs_event_set_wait() failed: %m");
```

Comment:
1) error is logged inside ucs_event_set_add. pls, remove
2) we should handle `UCS_INPROGRESS`

Reply:
What do you think this change?

```C
do {
    status = ucs_event_set_wait(thread->event_set,
                                UCS_ASYNC_EPOLL_MAX_EVENTS, timeout_ms,
                                ucs_async_thread_ev_handler, cb_arg,
                                &read_events);
} while (status == UCS_INPROGRESS);

if (status != UCS_OK) {
    ucs_error("ucs_event_set_wait() failed: %m");
}
```